json.go

```diff
@@ -0,0 +1,733 @@
+}
+
+void godror_dpiJsonNodeFree(dpiJsonNode *node)
+{
+    dpiJsonArray *array;
+    dpiJsonObject *obj;
+    uint32_t i;
+
+    if (node == NULL)
+    {
+        return;
+    }
+
+    switch (node->nativeTypeNum) {
+        case DPI_NATIVE_TYPE_BYTES:
+            if(node->value->asBytes.ptr) {
+                free(node->value->asBytes.ptr);
+                node->value->asBytes.ptr = NULL;
+            }
+            break;
+        case DPI_NATIVE_TYPE_JSON_ARRAY:
+            array = &node->value->asJsonArray;
+            if (array->elements) {
+                for (i = 0; i < array->numElements; i++) {
+                    if (array->elements[i].value) {
+                        godror_dpiJsonNodeFree(&array->elements[i]);
+                    }
+                }
+                free(array->elements);
+                array->elements = NULL;
+            }
+            if (array->elementValues) {
+                free(array->elementValues);
+                array->elementValues = NULL;
+            }
+            break;
+        case DPI_NATIVE_TYPE_JSON_OBJECT:
+            obj = &node->value->asJsonObject;
+            if (obj->fields) {
+                for (i = 0; i < obj->numFields; i++) {
+                    if (obj->fields[i].value)
+                        godror_dpiJsonNodeFree(&obj->fields[i]);
+                }
+                free(obj->fields);
+                obj->fields = NULL;
+            }
+            if (obj->fieldNames) {
+                for (i = 0; i < obj->numFields; i++) {
+                    if (obj->fieldNames[i]) {
+                        free(obj->fieldNames[i]);
+                        obj->fieldNames[i] = NULL;
+                    }
+                }
+                free(obj->fieldNames);
+                obj->fieldNames = NULL;
+            }
+            if (obj->fieldNameLengths) {
+                free(obj->fieldNameLengths);
+                obj->fieldNameLengths = NULL;
+            }
+            if (obj->fieldValues) {
+                free(obj->fieldValues);
+                obj->fieldValues = NULL;
+            }
+            break;
+    }
+}
+
+void godror_dpiJsonfreeMem(dpiJsonNode *node) {
+    if (node == NULL)
+    {
+        return;
+    }
+    godror_dpiJsonNodeFree(node);
+    godror_free_dpiNode(node);
+}
+
+*/
+import "C"
+
+import (
+	"encoding/json"
+	"errors"
+	"fmt"
+	"reflect"
+	"time"
+	"unsafe"
+)
+
+// JSONOption provides an option to retrieve scalar values
+// from JSON tree.
+//
+// DPI_JSON_OPT_NUMBER_AS_STRING - returns value stored as NUMBER in DB
+// as godror.Number.
+// DPI_JSON_OPT_DEAFULT - returns value stored as NUMBER in DB as float64.
+type JSONOption uint8
+
+var ErrInvalidJSON = errors.New("Invalid JSON Document")
+var ErrInvalidType = errors.New("Invalid JSON Scalar Type")
+
+const (
+	JSONOptDefault        = JSONOption(C.DPI_JSON_OPT_DEFAULT)
+	JSONOptNumberAsString = JSONOption(C.DPI_JSON_OPT_NUMBER_AS_STRING)
+)
+
+var NullJSONObject = JSONObject{}
```

Comment:
Why are these needed at all? "NullJSONObject" is longer than "JSONObject{}"!

But at least, unexport (start with lower case letter) them!

As it is know, anybody can accidentally modify them, and all functions that return a NullJSONObject,
may return something else!

Reply:
Removed the exported symbols. returned respective structs.